### PR TITLE
Added file upload functionality for freestyle jobs

### PR DESCRIPTION
**Pull request by vikulin**
Added file upload functionality in job configuration.

### CODE REVIEW

File: src/main/java/jenkins/plugins/slack/ActiveNotifier.java [hunk 1]
```diff
@@ -128,6 +128,9 @@ public void completed(AbstractBuild r) {
                     message = message + "\n" + getCommitList(r);
                 }
                 slackFactory.apply(r).publish(message, getBuildColor(r));
+                if(notifier.getUploadFiles()) {
```
Comment (timja) [inline src/main/java/jenkins/plugins/slack/ActiveNotifier.java:131]:
```suggestion
                if (notifier.getUploadFiles()) {
```

File: src/main/java/jenkins/plugins/slack/SlackNotifier.java [hunk 2]
```diff
@@ -25,6 +24,7 @@
 import java.util.logging.Level;
 import java.util.logging.Logger;
 import java.util.stream.Stream;
+import javax.annotation.CheckForNull;
```
Comment (timja) [inline src/main/java/jenkins/plugins/slack/SlackNotifier.java:27]: please use `edu.umd.cs.findbugs.annotations.CheckForNull`, the javax ones have a licensing issue.

File: src/main/java/jenkins/plugins/slack/SlackNotifier.java [hunk 3]
```diff
@@ -409,27 +429,27 @@ public SlackNotifier(CommitInfoChoice commitInfoChoice) {
     public SlackNotifier(final String baseUrl, final String teamDomain, final String authToken, final boolean botUser, final String room, final String tokenCredentialId,
                          final String sendAs, final boolean startNotification, final boolean notifyAborted, final boolean notifyFailure,
                          final boolean notifyNotBuilt, final boolean notifySuccess, final boolean notifyUnstable, final boolean notifyRegression, final boolean notifyBackToNormal,
-                         final boolean notifyRepeatedFailure, final boolean includeTestSummary, final boolean includeFailedTests,
+                         final boolean notifyRepeatedFailure, final boolean includeTestSummary, final boolean includeFailedTests, final boolean uploadFiles, final String artifactIncludes,
```
Comment (timja) [inline src/main/java/jenkins/plugins/slack/SlackNotifier.java:432]: the only constructor that needs updating is the builder, the rest are deprecated

File: src/main/java/jenkins/plugins/slack/StandardSlackService.java [hunk 4]
```diff
@@ -207,6 +214,43 @@ boolean postToSlack(String apiEndpoint, JSONObject body) {
...
+            for(String roomId : roomIds) {
+                SlackFileRequest slackFileRequest = new SlackFileRequest(
+                workspace, populatedToken, roomId, null, artifactIncludes);
+                try {
+                    workspace.getChannel().callAsync(new SlackUploadFileRunner(log, Jenkins.get().proxy, slackFileRequest)).get();
+                } catch (IllegalStateException e) {
+                    logger.log(Level.WARNING, "IllegalStateException", e);
+                    result = false;
+                } catch (InterruptedException e) {
+                    logger.log(Level.WARNING, "InterruptedException", e);
```
Comment (timja) [inline src/main/java/jenkins/plugins/slack/StandardSlackService.java:237]: No need to log all of these separately, the exception message / stacktrace will say what it is, can you collapse it in to one catch block please

File: src/main/java/jenkins/plugins/slack/pipeline/SlackUploadFileRunner.java [hunk 5]
```diff
@@ -24,7 +26,7 @@
 import org.apache.http.util.EntityUtils;
 import org.json.JSONObject;
 
-class SlackUploadFileRunner extends MasterToSlaveCallable<Boolean, Throwable> implements Serializable {
+public class SlackUploadFileRunner extends MasterToSlaveCallable<Boolean, Throwable> implements Serializable {
```
Comment (timja) [inline src/main/java/jenkins/plugins/slack/pipeline/SlackUploadFileRunner.java:29]: Let's keep it out of the public API for now

```suggestion
@Restricted(NoExternalUse.class)
public class SlackUploadFileRunner extends MasterToSlaveCallable<Boolean, Throwable> implements Serializable {
```

File: src/main/java/jenkins/plugins/slack/pipeline/SlackUploadFileRunner.java [hunk 6]
```diff
@@ -87,53 +98,49 @@ private boolean doIt(File file) {
...
-
-            ResponseHandler<JSONObject> responseHandler = response -> {
-                int status = response.getStatusLine().getStatusCode();
-                if (status >= 200 && status < 300) {
-                    HttpEntity entity = response.getEntity();
-                    return entity != null ? new org.json.JSONObject(EntityUtils.toString(entity)) : null;
-                } else {
-                    logger.log(Level.WARNING, UPLOAD_FAILED_TEMPLATE + status);
-                    return null;
+            for(File file:files) {
```
Comment (timja) [inline src/main/java/jenkins/plugins/slack/pipeline/SlackUploadFileRunner.java:101]:
```suggestion
            for (File file:files) {
```

File: src/main/java/jenkins/plugins/slack/pipeline/SlackUploadFileRunner.java [hunk 7]
```diff
@@ -87,53 +98,49 @@ private boolean doIt(File file) {
...
 
-            if (responseBody != null && !responseBody.getBoolean("ok")) {
-                listener.getLogger().println(UPLOAD_FAILED_TEMPLATE + responseBody.toString());
-                return false;
-            } else {
-                return true;
+                HttpUriRequest request = RequestBuilder
+                        .post(API_URL)
+                        .setEntity(multipartEntityBuilder.build())
+                        .addHeader("Authorization", "Bearer "+token)
```
Comment (timja) [inline src/main/java/jenkins/plugins/slack/pipeline/SlackUploadFileRunner.java:120]:
```suggestion
                        .addHeader("Authorization", "Bearer " + token)
```

File: src/main/resources/jenkins/plugins/slack/SlackNotifier/help-artifactIncludes.html [hunk 8]
```diff
@@ -0,0 +1 @@
+ setting that specifies the generated raw XML report files, such as ‘myproject/target/test-reports/*.xml’. Basedir of the fileset is <a href="http://ant.apache.org/manual/Types/fileset.html">Fileset ‘includes’ </a><a href="ws/">the workspace root</a>.
```
Comment (timja) [inline src/main/resources/jenkins/plugins/slack/SlackNotifier/help-artifactIncludes.html:1]: this doesn't really make sense to me, can you rephrase it possibly?

File: src/main/java/jenkins/plugins/slack/workflow/SlackSendStep.java [hunk 9]
```diff
@@ -47,6 +47,8 @@
     private static final Logger logger = Logger.getLogger(SlackSendStep.class.getName());
 
     private String message;
+    private boolean uploadFiles;
```
Comment (timja) [inline src/main/java/jenkins/plugins/slack/workflow/SlackSendStep.java:50]: why have these been added here? there's a separate step for pipeline `slackUploadFile`

**Comment by vikulin** [comment]
> 
> 
> Sorry for the delay I missed this,
> 
> A few initial comments, I haven't tested it out yet.
> 
> Could you expand the justification for the change please?
> This plugin already has file upload functionality, from what I can tell you've done this for freestyle builds? You also added some changes to the slackSendStep possibly to add support there but I'm not sure why?
> 
> Also your build is failing for javadoc errors

Yes, I have done it for free style jobs which has no file uploading available in UI.

File: src/main/java/jenkins/plugins/slack/SlackNotifier.java (hunk 2 above)
Comment (vikulin) [reply src/main/java/jenkins/plugins/slack/SlackNotifier.java:27]: I followed the comment from the class which refers that the annotation is deprecated:
```
/**
 * The annotated element might be null, and uses of the element should check for
 * null.
 *
 * When this annotation is applied to a method it applies to the method return
 * value.
 *
 * @deprecated - use {@link javax.annotation.CheckForNull} instead.
 **/
@Documented
@Target({ ElementType.FIELD, ElementType.METHOD, ElementType.PARAMETER, ElementType.LOCAL_VARIABLE })
@Retention(RetentionPolicy.CLASS)
@javax.annotation.Nonnull(when = When.MAYBE)
@TypeQualifierNickname
@Deprecated
public @interface CheckForNull {

}
```

File: src/main/java/jenkins/plugins/slack/SlackNotifier.java (hunk 3 above)
Comment (vikulin) [reply src/main/java/jenkins/plugins/slack/SlackNotifier.java:432]: All public constructors are deprecated. Do I need to remove new fields from all of them?

File: src/main/resources/jenkins/plugins/slack/SlackNotifier/help-artifactIncludes.html (hunk 8 above)
Comment (vikulin) [reply src/main/resources/jenkins/plugins/slack/SlackNotifier/help-artifactIncludes.html]: We still need some help for the upload field.

**Comment by vikulin** [comment]
Completed several changes.

**Comment by vikulin** [comment]
@timja, all changes have been made. Please review it.

**Comment by vikulin** [comment]
@timja , upload test has been done successfully.

**Review by timja — APPROVED** [review]

**Comment by timja** [comment]
Can you take a look at the build failure, something to do with javadoc, not sure exactly which one has caused it to fail though

**Comment by vikulin** [comment]
Javadoc fix

**Comment by vikulin** [comment]
Second Javadoc fix

**Comment by vikulin** [comment]
@timja , all errors have been fixed

**Review by timja — APPROVED** [review]
javadoc isn't that great but can be improved later
